File: emarsys-sdk/src/main/java/com/emarsys/Emarsys.java
package com.emarsys;

import android.app.Activity;
import android.content.Intent;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.emarsys.config.ConfigApi;
import com.emarsys.config.EmarsysConfig;
import com.emarsys.core.RunnerProxy;
import com.emarsys.core.api.experimental.FlipperFeature;
import com.emarsys.core.api.result.CompletionListener;
import com.emarsys.core.api.result.ResultListener;
import com.emarsys.core.api.result.Try;
import com.emarsys.core.database.DatabaseContract;
import com.emarsys.core.database.trigger.TriggerEvent;
import com.emarsys.core.database.trigger.TriggerType;
import com.emarsys.core.device.DeviceInfo;
import com.emarsys.core.di.DependencyInjection;
import com.emarsys.core.feature.FeatureRegistry;
import com.emarsys.core.util.Assert;
import com.emarsys.di.DefaultEmarsysDependencyContainer;
import com.emarsys.di.EmarsysDependencyInjection;
import com.emarsys.di.EmarysDependencyContainer;
import com.emarsys.inapp.InAppApi;
import com.emarsys.inbox.InboxApi;
import com.emarsys.mobileengage.api.EventHandler;
import com.emarsys.mobileengage.api.inbox.Notification;
import com.emarsys.mobileengage.api.inbox.NotificationInboxStatus;
import com.emarsys.predict.PredictApi;
import com.emarsys.predict.api.model.CartItem;
import com.emarsys.predict.api.model.Logic;
import com.emarsys.predict.api.model.Product;
import com.emarsys.predict.api.model.RecommendationFilter;
import com.emarsys.push.PushApi;

import java.util.List;
import java.util.Map;

import static com.emarsys.feature.InnerFeature.MOBILE_ENGAGE;
import static com.emarsys.feature.InnerFeature.PREDICT;

public class Emarsys {

    private static ConfigApi config;

    public static void setup(@NonNull EmarsysConfig emarsysConfig) {
        Assert.notNull(emarsysConfig, "Config must not be null!");

        for (FlipperFeature feature : emarsysConfig.getExperimentalFeatures()) {
            FeatureRegistry.enableFeature(feature);
        }

        DependencyInjection.setup(new DefaultEmarsysDependencyContainer(emarsysConfig));

        config = getContainer().getConfig();

        initializeApplicationCode(emarsysConfig);
        initializeInAppInternal(emarsysConfig);

        registerWatchDogs(emarsysConfig);

        registerDatabaseTriggers();

        initializeContact();
    }

    public static void setContact(@NonNull final String contactId) {
        getRunnerProxy().logException(new Runnable() {
            @Override
            public void run() {
                Assert.notNull(contactId, "ContactId must not be null!");

                if (FeatureRegistry.isFeatureEnabled(MOBILE_ENGAGE) ||
                        (!FeatureRegistry.isFeatureEnabled(MOBILE_ENGAGE) && !FeatureRegistry.isFeatureEnabled(PREDICT))) {
                    EmarsysDependencyInjection.mobileEngageInternal().setContact(contactId, null);
                }
                if (FeatureRegistry.isFeatureEnabled(PREDICT)) {
                    EmarsysDependencyInjection.predictInternal().setContact(contactId);
                }
            }
        });

    }

    public static void setContact(
            @NonNull final String contactId,
            @NonNull final CompletionListener completionListener) {
        getRunnerProxy().logException(new Runnable() {
            @Override
            public void run() {
                Assert.notNull(contactId, "ContactId must not be null!");
                Assert.notNull(completionListener, "CompletionListener must not be null!");

                if (FeatureRegistry.isFeatureEnabled(MOBILE_ENGAGE) ||
                        (!FeatureRegistry.isFeatureEnabled(MOBILE_ENGAGE) && !FeatureRegistry.isFeatureEnabled(PREDICT))) {
                    EmarsysDependencyInjection.mobileEngageInternal().setContact(contactId, completionListener);
                }
                if (FeatureRegistry.isFeatureEnabled(PREDICT)) {
                    EmarsysDependencyInjection.predictInternal().setContact(contactId);
                }
            }
        });
    }

    public static void clearContact() {
        getRunnerProxy().logException(new Runnable() {
            @Override
            public void run() {
                if (FeatureRegistry.isFeatureEnabled(MOBILE_ENGAGE) ||
                        (!FeatureRegistry.isFeatureEnabled(MOBILE_ENGAGE) && !FeatureRegistry.isFeatureEnabled(PREDICT))) {
                    EmarsysDependencyInjection.mobileEngageInternal().clearContact(null);
                }
                if (FeatureRegistry.isFeatureEnabled(PREDICT)) {
                    EmarsysDependencyInjection.predictInternal().clearContact();
                }
            }
        });
    }

    public static void clearContact(@NonNull final CompletionListener completionListener) {
        getRunnerProxy().logException(new Runnable() {
            @Override
            public void run() {
                Assert.notNull(completionListener, "CompletionListener must not be null!");

                if (FeatureRegistry.isFeatureEnabled(MOBILE_ENGAGE) ||
                        (!FeatureRegistry.isFeatureEnabled(MOBILE_ENGAGE) && !FeatureRegistry.isFeatureEnabled(PREDICT))) {
                    EmarsysDependencyInjection.mobileEngageInternal().clearContact(completionListener);
                }
                if (FeatureRegistry.isFeatureEnabled(PREDICT)) {
                    EmarsysDependencyInjection.predictInternal().clearContact();
                }
            }
        });

    }

    public static void trackDeepLink(@NonNull final Activity activity,
                                     @NonNull final Intent intent) {
        getRunnerProxy().logException(new Runnable() {
            @Override
            public void run() {
                Assert.notNull(activity, "Activity must not be null!");
                Assert.notNull(intent, "Intent must not be null!");

                EmarsysDependencyInjection.deepLinkInternal().trackDeepLinkOpen(activity, intent, null);
            }
        });
    }

    public static void trackDeepLink(@NonNull final Activity activity,
                                     @NonNull final Intent intent,
                                     @NonNull final CompletionListener completionListener) {
        getRunnerProxy().logException(new Runnable() {
            @Override
            public void run() {
                Assert.notNull(activity, "Activity must not be null!");
                Assert.notNull(intent, "Intent must not be null!");
                Assert.notNull(completionListener, "CompletionListener must not be null!");

                EmarsysDependencyInjection.deepLinkInternal().trackDeepLinkOpen(activity, intent, completionListener);
            }
        });
    }

    public static void trackCustomEvent(
            @NonNull final String eventName,
            @Nullable final Map<String, String> eventAttributes) {
        getRunnerProxy().logException(new Runnable() {
            @Override
            public void run() {
                Assert.notNull(eventName, "EventName must not be null!");

                EmarsysDependencyInjection.eventServiceInternal().trackCustomEvent(eventName, eventAttributes, null);
            }
        });
    }

    public static void trackCustomEvent(
            @NonNull final String eventName,
            @Nullable final Map<String, String> eventAttributes,
            @NonNull final CompletionListener completionListener) {
        getRunnerProxy().logException(new Runnable() {
            @Override
            public void run() {
                Assert.notNull(eventName, "EventName must not be null!");
                Assert.notNull(completionListener, "CompletionListener must not be null!");

                EmarsysDependencyInjection.eventServiceInternal().trackCustomEvent(eventName, eventAttributes, completionListener);
            }
        });
    }

    static PushApi getPush() {
        return EmarsysDependencyInjection.push();
    }

    static InAppApi getInApp() {
        return EmarsysDependencyInjection.inApp();
    }

    static InboxApi getInbox() {
        return EmarsysDependencyInjection.inbox();
    }

    static PredictApi getPredict() {
        return EmarsysDependencyInjection.predict();
    }

    public static class Config {

        @NonNull
        public static int getContactFieldId() {
            return config.getContactFieldId();
        }

        public static void changeApplicationCode(@Nullable String applicationCode) {
            config.changeApplicationCode(applicationCode, null);
        }

        public static void changeApplicationCode(@Nullable String applicationCode, @NonNull CompletionListener completionListener) {
            config.changeApplicationCode(applicationCode, completionListener);
        }

        @Nullable
        public static String getApplicationCode() {
            return config.getApplicationCode();
        }

        public static void changeMerchantId(@Nullable String merchantId) {
            config.changeMerchantId(merchantId);
        }

        @Nullable
        public static String getMerchantId() {
            return config.getMerchantId();
        }

    }

    public static class Push {

        public static void trackMessageOpen(@NonNull final Intent intent) {
            getPush().trackMessageOpen(intent);
        }

        public static void trackMessageOpen(
                @NonNull final Intent intent,
                @NonNull final CompletionListener completionListener) {
            getPush().trackMessageOpen(intent, completionListener);
        }

        public static void setPushToken(@NonNull final String pushToken) {
            getPush().setPushToken(pushToken);
        }

        public static void setPushToken(
                @NonNull final String pushToken,
                @NonNull final CompletionListener completionListener) {
            getPush().setPushToken(pushToken, completionListener);
        }

        public static void clearPushToken() {
            getPush().clearPushToken();
        }

        public static void clearPushToken(final CompletionListener completionListener) {
            getPush().clearPushToken(completionListener);
        }
    }

    public static class Predict {

        public static void trackCart(@NonNull final List<CartItem> items) {
            getPredict().trackCart(items);
        }

        public static void trackPurchase(@NonNull final String orderId,
                                         @NonNull final List<CartItem> items) {
            getPredict().trackPurchase(orderId, items);
        }

        public static void trackItemView(@NonNull final String itemId) {
            getPredict().trackItemView(itemId);
        }

        public static void trackCategoryView(@NonNull final String categoryPath) {
            getPredict().trackCategoryView(categoryPath);
        }

        public static void trackSearchTerm(@NonNull final String searchTerm) {
            getPredict().trackSearchTerm(searchTerm);
        }

        public static void trackTag(@NonNull String tag, @Nullable Map<String, String> attributes) {
            getPredict().trackTag(tag, attributes);
        }

        public static void recommendProducts(@NonNull final Logic recommendationLogic, @NonNull final ResultListener<Try<List<Product>>> resultListener) {
            getPredict().recommendProducts(recommendationLogic, resultListener);
        }

        public static void recommendProducts(@NonNull final Logic recommendationLogic, @NonNull final Integer limit, @NonNull final ResultListener<Try<List<Product>>> resultListener) {
            getPredict().recommendProducts(recommendationLogic, limit, resultListener);
        }

        public static void recommendProducts(@NonNull final Logic recommendationLogic, @NonNull final List<RecommendationFilter> recommendationFilters, @NonNull ResultListener<Try<List<Product>>> resultListener) {
            getPredict().recommendProducts(recommendationLogic, recommendationFilters, resultListener);
        }

        public static void recommendProducts(@NonNull final Logic recommendationLogic, @NonNull final List<RecommendationFilter> recommendationFilters, @NonNull final Integer limit, @NonNull ResultListener<Try<List<Product>>> resultListener) {
            getPredict().recommendProducts(recommendationLogic, limit, recommendationFilters, resultListener);
        }

        public static void trackRecommendationClick(@NonNull final Product product) {
            getPredict().trackRecommendationClick(product);
        }
    }

    public static class InApp {

        public static void pause() {
            getInApp().pause();
        }

        public static void resume() {
            getInApp().resume();
        }

        public static boolean isPaused() {
            return getInApp().isPaused();
        }

        public static void setEventHandler(@NonNull EventHandler eventHandler) {
            getInApp().setEventHandler(eventHandler);
        }
    }

    public static class Inbox {

        public static void fetchNotifications(@NonNull ResultListener<Try<NotificationInboxStatus>> resultListener) {
            getInbox().fetchNotifications(resultListener);
        }


        public static void trackNotificationOpen(@NonNull Notification notification) {
            getInbox().trackNotificationOpen(notification);
        }

        public static void trackNotificationOpen(@NonNull Notification notification, @NonNull CompletionListener completionListener) {
            getInbox().trackNotificationOpen(notification, completionListener);
        }

        public static void resetBadgeCount() {
            getInbox().resetBadgeCount();
        }

        public static void resetBadgeCount(@NonNull CompletionListener completionListener) {
            getInbox().resetBadgeCount(completionListener);
        }
    }

    private static EmarysDependencyContainer getContainer() {
        return DependencyInjection.getContainer();
    }

    private static RunnerProxy getRunnerProxy() {
        return getContainer().getRunnerProxy();
    }

    private static void initializeInAppInternal(@NonNull EmarsysConfig config) {
        EventHandler inAppEventHandler = config.getInAppEventHandler();

        if (inAppEventHandler != null) {
            getInApp().setEventHandler(inAppEventHandler);
        }
    }

    private static void registerWatchDogs(EmarsysConfig config) {
        config.getApplication().registerActivityLifecycleCallbacks(getContainer().getActivityLifecycleWatchdog());
        config.getApplication().registerActivityLifecycleCallbacks(getContainer().getCurrentActivityWatchdog());
    }

    private static void registerDatabaseTriggers() {
        if (FeatureRegistry.isFeatureEnabled(PREDICT)) {
            getContainer().getCoreSQLiteDatabase().registerTrigger(
                    DatabaseContract.SHARD_TABLE_NAME,
                    TriggerType.AFTER,
                    TriggerEvent.INSERT,
                    getContainer().getPredictShardTrigger());
        }

        getContainer().getCoreSQLiteDatabase().registerTrigger(
                DatabaseContract.SHARD_TABLE_NAME,
                TriggerType.AFTER,
                TriggerEvent.INSERT,
                getContainer().getLogShardTrigger());
    }

    private static void initializeContact() {
        Integer deviceInfoHash = getContainer().getDeviceInfoHashStorage().get();
        String contactToken = getContainer().getContactTokenStorage().get();
        String contactFieldValue = getContainer().getContactFieldValueStorage().get();
        String clientState = getContainer().getClientStateStorage().get();
        DeviceInfo deviceInfo = getContainer().getDeviceInfo();

        if (contactToken == null && contactFieldValue == null) {
            if (clientState == null || deviceInfoHash != null && !deviceInfoHash.equals(deviceInfo.getHash())) {
                EmarsysDependencyInjection.clientServiceInternal().trackDeviceInfo();
            }
            EmarsysDependencyInjection.mobileEngageInternal().setContact(null, null);
        }
    }

    private static void initializeApplicationCode(EmarsysConfig emarsysConfig) {
        getContainer().getRequestContext().getApplicationCodeStorage().set(emarsysConfig.getMobileEngageApplicationCode());
    }
}

File: emarsys-sdk/src/main/java/com/emarsys/Emarsys.kt
package com.emarsys

import android.app.Activity
import android.content.Intent
import android.util.Log
import androidx.lifecycle.ProcessLifecycleOwner
import com.emarsys.common.feature.InnerFeature.EVENT_SERVICE_V4
import com.emarsys.common.feature.InnerFeature.MOBILE_ENGAGE
import com.emarsys.common.feature.InnerFeature.PREDICT
import com.emarsys.config.ConfigApi
import com.emarsys.config.EmarsysConfig
import com.emarsys.core.api.proxyApi
import com.emarsys.core.api.proxyWithLogExceptions
import com.emarsys.core.api.result.CompletionListener
import com.emarsys.core.database.DatabaseContract
import com.emarsys.core.database.trigger.TriggerEvent
import com.emarsys.core.database.trigger.TriggerType
import com.emarsys.core.feature.FeatureRegistry
import com.emarsys.core.util.log.Logger
import com.emarsys.core.util.log.entry.CrashLog
import com.emarsys.core.util.log.entry.StatusLog
import com.emarsys.di.DefaultEmarsysDependencies
import com.emarsys.di.EmarsysDependencyInjection
import com.emarsys.di.emarsys
import com.emarsys.di.isEmarsysComponentSetup
import com.emarsys.geofence.GeofenceApi
import com.emarsys.inapp.InAppApi
import com.emarsys.inbox.MessageInboxApi
import com.emarsys.mobileengage.di.mobileEngage
import com.emarsys.oneventaction.OnEventActionApi
import com.emarsys.predict.PredictApi
import com.emarsys.push.PushApi

object Emarsys {

    @JvmStatic
    val geofence: GeofenceApi
        get() = EmarsysDependencyInjection.geofence()

    @JvmStatic
    val config: ConfigApi
        get() = emarsys().config

    @JvmStatic
    val push: PushApi
        get() = EmarsysDependencyInjection.push()

    @JvmStatic
    val inApp: InAppApi
        get() = EmarsysDependencyInjection.inApp()

    @JvmStatic
    val onEventAction: OnEventActionApi
        get() = EmarsysDependencyInjection.onEventAction()

    @JvmStatic
    val messageInbox: MessageInboxApi
        get() = EmarsysDependencyInjection.messageInbox()

    @JvmStatic
    val predict: PredictApi
        get() = EmarsysDependencyInjection.predict()

    @JvmStatic
    fun setup(emarsysConfig: EmarsysConfig) {
        for (feature in emarsysConfig.experimentalFeatures) {
            FeatureRegistry.enableFeature(feature)
        }

        if (emarsysConfig.applicationCode != null) {
            FeatureRegistry.enableFeature(MOBILE_ENGAGE)
            FeatureRegistry.enableFeature(EVENT_SERVICE_V4)
        }

        if (emarsysConfig.merchantId != null) {
            FeatureRegistry.enableFeature(PREDICT)
        }

        if (!isEmarsysComponentSetup()) {
            DefaultEmarsysDependencies(emarsysConfig)
        }

        emarsys().concurrentHandlerHolder.postOnMain {
            try {
                registerLifecycleObservers()
                registerWatchDogs(emarsysConfig)
            } catch (e: Throwable) {
                Logger.error(CrashLog(e))
            }
        }


        emarsys().concurrentHandlerHolder.coreHandler.post {
            registerDatabaseTriggers()

            if (FeatureRegistry.isFeatureEnabled(MOBILE_ENGAGE)) {
                initializeMobileEngageContact()
            }
        }

        refreshRemoteConfig(emarsysConfig.applicationCode)
    }

    private fun registerLifecycleObservers() {
        val appLifecycleObserver = emarsys().appLifecycleObserver
        ProcessLifecycleOwner.get().lifecycle.addObserver(appLifecycleObserver)
    }

    @JvmStatic
    @JvmOverloads
    fun setAuthenticatedContact(
        contactFieldId: Int,
        openIdToken: String,
        completionListener: CompletionListener? = null
    ) {
        if (FeatureRegistry.isFeatureEnabled(MOBILE_ENGAGE)
            || (!FeatureRegistry.isFeatureEnabled(MOBILE_ENGAGE)
                    && !FeatureRegistry.isFeatureEnabled(PREDICT))
        ) {
            EmarsysDependencyInjection.mobileEngageApi()
                .proxyApi(mobileEngage().concurrentHandlerHolder)
                .setAuthenticatedContact(contactFieldId, openIdToken, completionListener)
        }
    }

    @JvmStatic
    @JvmOverloads
    fun setContact(
        contactFieldId: Int,
        contactFieldValue: String,
        completionListener: CompletionListener? = null
    ) {
        executeMobileEngageOrPredictOnlyLogic(
            mobileEngageLogic = {
                setMobileEngageContact(
                    contactFieldId,
                    contactFieldValue,
                    completionListener
                )
            },
            predictOnlyLogic = {
                setPredictOnlyContact(
                    contactFieldId,
                    contactFieldValue,
                    completionListener
                )
            }
        )
    }

    private fun setMobileEngageContact(
        contactFieldId: Int,
        contactFieldValue: String,
        completionListener: CompletionListener?
    ) {
        EmarsysDependencyInjection.mobileEngageApi()
            .proxyApi(mobileEngage().concurrentHandlerHolder)
            .setContact(contactFieldId, contactFieldValue, completionListener)
    }

    private fun setPredictOnlyContact(
        contactFieldId: Int,
        contactFieldValue: String,
        completionListener: CompletionListener?
    ) {
        EmarsysDependencyInjection.predictRestrictedApi()
            .proxyApi(mobileEngage().concurrentHandlerHolder)
            .setContact(contactFieldId, contactFieldValue, completionListener)
    }

    @JvmStatic
    @JvmOverloads
    fun clearContact(completionListener: CompletionListener? = null) {
        executeMobileEngageOrPredictOnlyLogic(
            mobileEngageLogic = { clearMobileEngageContact(completionListener) },
            predictOnlyLogic = { clearPredictOnlyContact(completionListener) }
        )
    }

    private fun clearMobileEngageContact(completionListener: CompletionListener?) {
        EmarsysDependencyInjection.mobileEngageApi()
            .proxyApi(mobileEngage().concurrentHandlerHolder)
            .clearContact(completionListener)
        EmarsysDependencyInjection.predictRestrictedApi()
            .proxyApi(mobileEngage().concurrentHandlerHolder)
            .clearVisitorId()
    }

    private fun clearPredictOnlyContact(completionListener: CompletionListener?) {
        EmarsysDependencyInjection.predictRestrictedApi()
            .proxyApi(mobileEngage().concurrentHandlerHolder)
            .clearPredictOnlyContact(completionListener)
    }

    private fun executeMobileEngageOrPredictOnlyLogic(
        mobileEngageLogic: () -> Unit,
        predictOnlyLogic: () -> Unit
    ) {
        if (FeatureRegistry.isFeatureEnabled(MOBILE_ENGAGE)
            || (!FeatureRegistry.isFeatureEnabled(MOBILE_ENGAGE)
                    && !FeatureRegistry.isFeatureEnabled(PREDICT))
        ) {
            mobileEngageLogic()
        } else if (FeatureRegistry.isFeatureEnabled(PREDICT)) {
            predictOnlyLogic()
        }
    }

    @JvmStatic
    @JvmOverloads
    fun trackDeepLink(
        activity: Activity,
        intent: Intent,
        completionListener: CompletionListener? = null
    ) {
        EmarsysDependencyInjection.deepLinkApi()
            .proxyApi(mobileEngage().concurrentHandlerHolder)
            .trackDeepLinkOpen(activity, intent, completionListener)
    }

    @JvmStatic
    @JvmOverloads
    fun trackCustomEvent(
        eventName: String,
        eventAttributes: Map<String, String>?,
        completionListener: CompletionListener? = null
    ) {
        EmarsysDependencyInjection.eventServiceApi()
            .proxyApi(mobileEngage().concurrentHandlerHolder)
            .trackCustomEventAsync(eventName, eventAttributes, completionListener)
    }

    private fun registerWatchDogs(config: EmarsysConfig) {
        config.application.registerActivityLifecycleCallbacks(emarsys().currentActivityWatchdog)
        config.application.registerActivityLifecycleCallbacks(emarsys().activityLifecycleWatchdog)
        config.application.registerActivityLifecycleCallbacks(emarsys().transitionSafeCurrentActivityWatchdog)
    }

    private fun registerDatabaseTriggers() {
            emarsys().coreSQLiteDatabase
                .registerTrigger(
                    DatabaseContract.SHARD_TABLE_NAME,
                    TriggerType.AFTER,
                    TriggerEvent.INSERT,
                    emarsys().predictShardTrigger
                )

        emarsys().coreSQLiteDatabase
            .registerTrigger(
                DatabaseContract.SHARD_TABLE_NAME,
                TriggerType.AFTER,
                TriggerEvent.INSERT,
                emarsys().logShardTrigger
            )
    }

    private fun initializeMobileEngageContact() {
        val deviceInfoPayload = emarsys().deviceInfoPayloadStorage.get()
        val contactToken = emarsys().contactTokenStorage.get()
        val requestContext = emarsys().requestContext
        val clientState = emarsys().clientStateStorage.get()
        val deviceInfo = emarsys().deviceInfo

        if (contactToken == null && !requestContext.hasContactIdentification()) {
            if (clientState == null || deviceInfoPayload != null && deviceInfoPayload != deviceInfo.deviceInfoPayload) {
                EmarsysDependencyInjection.clientServiceApi()
                    .proxyWithLogExceptions()
                    .trackDeviceInfo(null)
            }
            EmarsysDependencyInjection.mobileEngageApi()
                .proxyWithLogExceptions()
                .setContact()
        }
    }

    private fun refreshRemoteConfig(applicationCode: String?) {
        if (!listOf(
                "",
                "null",
                "nil",
                "0"
            ).contains(applicationCode?.lowercase()) && applicationCode != null
        ) {
            emarsys().configInternal.proxyApi(mobileEngage().concurrentHandlerHolder)
                .refreshRemoteConfig {
                    it?.let {
                        val logEntry = StatusLog(
                            Emarsys::class.java,
                            "refreshRemoteConfig",
                            mapOf("applicationCode" to applicationCode, "exception" to it.message)
                        )
                        Logger.log(logEntry)
                    }
                }
        } else {
            Log.w("EmarsysSdk", "Invalid applicationCode: $applicationCode")
            val logEntry = StatusLog(
                Emarsys::class.java,
                "refreshRemoteConfig",
                mapOf("applicationCode" to applicationCode)
            )
            Logger.log(logEntry)
        }
    }
}